SoundBank

gnarf opened the pull request:
(just up for ease of communication, this is purely a first draft right now)

gnarf committed:
First Draft SoundBank + EffectChain

gnarf commented on src/SoundBank.js:
```diff
@@ -0,0 +1,84 @@
+
+    stop (target, soundId) {
+        if (this.playerTargets.get(soundId) === target) {
+            this.soundPlayers[soundId].stop();
```
should probaby use `this.getSound(soundId)` here for future refactor possibilities.

gnarf commented on src/SoundBank.js:
```diff
@@ -0,0 +1,84 @@
+        this.soundEffects.clear();
+        for (const soundId in this.soundPlayers) {
+            if (this.soundPlayers.hasOwnProperty(soundId)) {
+                this.soundPlayers[soundId].dispose();
```
I feel like this one can stay as is in the dispose loop, no need to use getSound

mzgoddard commented on src/SoundBank.js:
```diff
@@ -0,0 +1,84 @@
+    getSound (soundId) {
+        if (!this.soundPlayers[soundId]) {
+            this.soundPlayers[soundId] = new SoundPlayer(this.audioEngine, {
+                id: soundId, buffer: this.audioEngine.audioBuffers[soundId]
```
We should get rid of `audioBuffers`. Its a memory leak waiting to happen.

mzgoddard commented on src/SoundBank.js:
```diff
@@ -0,0 +1,84 @@
+
+class SoundBank {
+    constructor (audioEngine) {
+        this.audioEngine = audioEngine;
```
I think it'd be best if SoundBank does not know of AudioEngine.

gnarf replied:
The only reason it needs it is for `addSound` and knowing the output node to connect to I think...

mzgoddard commented on src/effects/EffectChain.js:
```diff
@@ -0,0 +1,74 @@
+    addSoundPlayer (soundPlayer) {
+        if (!this._soundPlayers.has(soundPlayer)) {
+            this._soundPlayers.add(soundPlayer);
+            this._effects.forEach(effect => effect.update());
```
I think making update a function to mirror Effect's API would be good.

mzgoddard commented on src/effects/EffectChain.js:
```diff
@@ -0,0 +1,74 @@
+class EffectChain {
+    constructor (audioEngine) {
```
Should we have an `effects` argument?

mzgoddard commented on src/effects/EffectChain.js:
```diff
@@ -0,0 +1,74 @@
+            } else if ('soundEffects' in target && effect.name in target.soundEffects) {
+                effect.set(target.soundEffects[effect.name]);
+            } else {
+                effect.set(effect.DEFAULT_VALUE);
```
This feels iffy.

gnarf replied:
agree - this whole bit didn't work out as nicely as I'd hoped on the VM side

mzgoddard commented on src/effects/EffectChain.js:
```diff
@@ -0,0 +1,74 @@
+        this._effects.reduceRight((nextNode, effect) => {
+            effect.connect(nextNode);
+            return effect;
+        }, this.audioEngine);
```
Lets make connecting to audioEngine done through a `.connect` method like Effect and SoundPlayer.

gnarf replied:
you mean that soundbank would connect soundplayer -> effectchain -> audioEngine ?

mzgoddard commented on src/SoundBank.js:
```diff
@@ -0,0 +1,84 @@
+
+    getSoundEffects (sound) {
+        if (!this.soundEffects.has(sound)) {
+            this.soundEffects.set(sound, new EffectsChain(this.audioEngine));
```
I think it'd be better to have a `this.effectsStandard.clone()` way to make new EffectsChains.

mzgoddard committed:
make EffectChain standalone

- EffectChain knows of audioEngine so it can create Effects
- EffectChain connects to the target specified by .connect(target)
- EffectChain can connect to other Effects or EffectChains
- EffectChain can clone itself and connect to the original's target
- Add EffectChain.update to mirror Effects API
- Use deepest setting first in setEffectsFromTarget

mzgoddard committed:
add SoundPlayers to SoundBank

- SoundBank may not create SoundPlayers (but it can call .take if need)
- All SoundPlayers SoundBank plays are given to it

mzgoddard committed:
pass main EffectChain to SoundBank from Engine.createBank

mzgoddard committed:
comment SoundBank and EffectChain

mzgoddard committed:
add AudioEngine.createEffectChain

gnarf committed:
rm AudioPlayer and SoundPlayer

gnarf committed:
mv GreenPlayer.js SoundPlayer.js

gnarf commented on src/SoundBank.js:
```diff
@@ -0,0 +1,156 @@
+        effects.setEffectsFromTarget(target);
+        effects.addSoundPlayer(player);
+
+        player.connect(effects);
```
need to play before connect

gnarf committed:
Better order of stop/connect/play

mzgoddard committed:
create Effects in EffectChain in given order

The first created Effect is closest to the input. The last is closest
to the output.

gnarf commented on src/effects/EffectChain.js:
```diff
@@ -0,0 +1,182 @@
+
+        this.target = target;
+
+        firstEffect.connect(target);
```
is this backwards now?
